fs-storage/src/folder_storage.rs

```diff
@@ -0,0 +1,336 @@
+                    .unwrap()
+                    .to_str()
+                    .unwrap()
+                    .parse::<K>()
+                    .map_err(|_| {
+                        ArklibError::Storage(
+                            self.label.clone(),
+                            "Failed to parse key from filename".to_owned(),
+                        )
+                    })?;
+
+                let mut file = File::open(&path)?;
+                let mut buffer = Vec::new();
+                file.read_to_end(&mut buffer)?;
+
+                let value: V = bincode::deserialize(&buffer).map_err(|e| {
+                    ArklibError::Storage(
+                        self.label.clone(),
+                        format!("Failed to deserialize value: {}", e),
+                    )
+                })?;
+                data.entries.insert(key, value);
+            }
+        }
+        Ok(data)
+    }
+}
+
+impl<K, V> BaseStorage<K, V> for FolderStorage<K, V>
+where
+    K: Ord
+        + Clone
+        + serde::Serialize
+        + serde::de::DeserializeOwned
+        + std::str::FromStr
+        + std::fmt::Display,
+    V: Clone
+        + serde::Serialize
+        + serde::de::DeserializeOwned
+        + std::str::FromStr
+        + Monoid<V>,
+{
+    /// Set a key-value pair in the internal mapping
+    fn set(&mut self, key: K, value: V) {
+        self.data.entries.insert(key.clone(), value);
+        self.ram_timestamps.insert(key, SystemTime::now());
+    }
+
+    /// Remove an entry from the internal mapping given a key
+    fn remove(&mut self, id: &K) -> Result<()> {
+        self.data.entries.remove(id).ok_or_else(|| {
+            ArklibError::Storage(self.label.clone(), "Key not found".to_owned())
+        })?;
+        // self.ram_timestamps.remove(id);
+        // OR
+        self.ram_timestamps
+            .insert(id.clone(), SystemTime::now());
+        Ok(())
+    }
+
+    /// Compare the timestamp of the storage file
+    /// with the timestamp of the in-memory storage and the last written
+    /// to time to determine if either of the two requires syncing.
+    fn sync_status(&self) -> Result<SyncStatus> {
+        unimplemented!()
+    }
+
+    /// Sync the in-memory storage with the storage on disk
+    fn sync(&mut self) -> Result<()> {
+        match self.sync_status()? {
+            SyncStatus::InSync => Ok(()),
+            SyncStatus::MappingStale => self.read_fs().map(|_| ()),
+            SyncStatus::StorageStale => self.write_fs().map(|_| ()),
+            SyncStatus::Diverge => {
+                let data = self.load_fs_data()?;
+                self.merge_from(&data)?;
+                self.write_fs()?;
+                Ok(())
+            }
+        }
+    }
+
+    /// Read the data from file
+    fn read_fs(&mut self) -> Result<&BTreeMap<K, V>> {
+        let data = self.load_fs_data()?;
+        self.data = data;
+        self.disk_timestamps.clear();
+        for key in self.data.entries.keys() {
```

Comment:
Instead of collecting the disk timestamps separately. Why not collect them inside `load_fs_data` itself? Where you are already reading the data from the files you can also read the metadata.